### PR TITLE
1.20.5 and 1.20.6 support

### PR DESCRIPTION
**Pull request by Lagggpixel**

### CODE REVIEW

**Comment by tomkeuper** [comment]
The NBT implementation is completely different on 1_20_R5 so this would require a complete rewrite of the nbt implementation.

I have done some research and found a library that would handle all the NBT related stuff which would make updating easier aswell in the future. (https://docs.saicone.com/rtag/)
I think we should consider using that library and shade it into BW2023 (https://docs.saicone.com/rtag/#shaded-dependency)

File: versionsupport_v1_20_r3/src/main/java/com/tomkeuper/bedwars/support/version/v1_20_R3/v1_20_R3.java [hunk 1]
```diff
@@ -369,12 +372,12 @@ public void registerTntWhitelist(float endStoneBlast, float glassBlast) {
...
 
             var coloredGlass = new net.minecraft.world.level.block.Block[]{
-                    Blocks.ej, Blocks.ek, Blocks.el, Blocks.em,
-                    Blocks.en, Blocks.eo, Blocks.ep, Blocks.eq,
-                    Blocks.er, Blocks.es, Blocks.et, Blocks.eu,
-                    Blocks.ev, Blocks.ew, Blocks.ex, Blocks.ey,
+                    Blocks.ei, Blocks.ej, Blocks.ek, Blocks.el,
+                    Blocks.em, Blocks.en, Blocks.eo, Blocks.ep,
+                    Blocks.eq, Blocks.er, Blocks.es, Blocks.et,
+                    Blocks.eu, Blocks.ev, Blocks.ew, Blocks.ex,
```
Comment (tomkeuper) [inline versionsupport_v1_20_r3/src/main/java/com/tomkeuper/bedwars/support/version/v1_20_R3/v1_20_R3.java:380]: Why did you change these?

Comment (Lagggpixel) [reply versionsupport_v1_20_r3/src/main/java/com/tomkeuper/bedwars/support/version/v1_20_R3/v1_20_R3.java:380]: Those are the glass objects, previous ones were wrong.

File: versionsupport_v1_20_r5/src/main/java/com/tomkeuper/bedwars/support/version/v1_20_R5/v1_20_R5.java [hunk 2]
```diff
@@ -0,0 +1,889 @@
...
+        return item.getItem();
+    }
+
+    @Override
+    public org.bukkit.inventory.ItemStack getPlayerHead(Player player, org.bukkit.inventory.ItemStack copyTagFrom) {
+        org.bukkit.inventory.ItemStack head = new org.bukkit.inventory.ItemStack(materialPlayerHead());
+
+        if (copyTagFrom != null) {
+            var tag = getTag(copyTagFrom);
+            head = applyTag(head, tag);
```
Comment (Lagggpixel) [inline versionsupport_v1_20_r5/src/main/java/com/tomkeuper/bedwars/support/version/v1_20_R5/v1_20_R5.java]: I think this is wrong still. Not sure how to fix this one.

File: versionsupport_v1_20_r5/src/main/java/com/tomkeuper/bedwars/support/version/v1_20_R5/v1_20_R5.java [hunk 3]
```diff
@@ -0,0 +1,889 @@
...
+        return i.F();
+    }
+
+    private @Nullable NBTTagCompound getTag(@NotNull org.bukkit.inventory.ItemStack itemStack) {
+        RtagItem rtagItem = new RtagItem(itemStack);
+        OptionalType nbt = rtagItem.getOptional(VersionSupport.PLUGIN_TAG_GENERIC_KEY);
+        return (nbt.isEmpty() || nbt.isNotInstance(NBTTagCompound.class)) ? null : nbt.value();
+    }
+
+    public org.bukkit.inventory.ItemStack applyTag(org.bukkit.inventory.ItemStack itemStack, NBTTagCompound tag) {
```
Comment (Lagggpixel) [inline versionsupport_v1_20_r5/src/main/java/com/tomkeuper/bedwars/support/version/v1_20_R5/v1_20_R5.java]: This method can probably be removed.

Comment (Lagggpixel) [reply versionsupport_v1_20_r5/src/main/java/com/tomkeuper/bedwars/support/version/v1_20_R5/v1_20_R5.java]: There is one usage of it across the entire class.

File: versionsupport_v1_20_r5/src/main/java/com/tomkeuper/bedwars/support/version/v1_20_R5/v1_20_R5.java [hunk 4]
```diff
@@ -0,0 +1,889 @@
...
+            if (p.getInventory().getItemInOffHand().equals(i)) {
+                p.getInventory().setItemInOffHand(null);
+            } else {
+                p.getInventory().removeItem(i);
+            }
+            return;
+        }
+        i.setAmount(i.getAmount() - amount);
+        //noinspection UnstableApiUsage
+        p.updateInventory(); //TODO might be deprecated. Currently marked as unstable. Find a replacement later on.
```
Comment (Lagggpixel) [inline versionsupport_v1_20_r5/src/main/java/com/tomkeuper/bedwars/support/version/v1_20_R5/v1_20_R5.java]: I don't think this is needed anymore on the latest versions, inventory should update automatically according to Spigot.

File: versionsupport_v1_20_r5/src/main/java/com/tomkeuper/bedwars/support/version/v1_20_R5/v1_20_R5.java [hunk 5]
```diff
@@ -0,0 +1,889 @@
...
+            ex.printStackTrace();
+        }
+    }
+
+    @Override
+    public void voidKill(Player p) {
+        @SuppressWarnings("UnstableApiUsage")
+        EntityDamageEvent event = new EntityDamageEvent(p, EntityDamageEvent.DamageCause.VOID, DamageSource.builder(DamageType.GENERIC).build(), 1000.0);
+        //noinspection removal
+        p.setLastDamageCause(event); // TODO is deprecated and marked for removal. Find a replacement as soon as possible
```
Comment (Lagggpixel) [inline versionsupport_v1_20_r5/src/main/java/com/tomkeuper/bedwars/support/version/v1_20_R5/v1_20_R5.java]: This should be changed if possible, I've changed the EntityDamageEvent to make it not deprecated.

File: versionsupport_v1_20_r3/src/main/java/com/tomkeuper/bedwars/support/version/v1_20_R3/v1_20_R3.java (hunk 1 above)
Comment (tomkeuper) [reply versionsupport_v1_20_r3/src/main/java/com/tomkeuper/bedwars/support/version/v1_20_R3/v1_20_R3.java:380]: Ahh I see, good catch!

File: versionsupport_v1_20_r5/src/main/java/com/tomkeuper/bedwars/support/version/v1_20_R5/v1_20_R5.java (hunk 3 above)
Comment (tomkeuper) [reply versionsupport_v1_20_r5/src/main/java/com/tomkeuper/bedwars/support/version/v1_20_R5/v1_20_R5.java]: moved code inline

File: versionsupport_v1_20_r5/src/main/java/com/tomkeuper/bedwars/support/version/v1_20_R5/v1_20_R5.java (hunk 2 above)
Comment (tomkeuper) [reply versionsupport_v1_20_r5/src/main/java/com/tomkeuper/bedwars/support/version/v1_20_R5/v1_20_R5.java]: What do you think is wrong about it? The code just gets the NBT tag from the old item stack and moves it to the new one.

File: versionsupport_v1_20_r5/src/main/java/com/tomkeuper/bedwars/support/version/v1_20_R5/v1_20_R5.java (hunk 5 above)
Comment (tomkeuper) [reply versionsupport_v1_20_r5/src/main/java/com/tomkeuper/bedwars/support/version/v1_20_R5/v1_20_R5.java]: since NMS is version specific it doesnt really matter that it is marked for removal. (it was already marked deprecated in v1_20_R3.) It should only change when a new version is released.

File: versionsupport_v1_20_r5/src/main/java/com/tomkeuper/bedwars/support/version/v1_20_R5/v1_20_R5.java (hunk 4 above)
Comment (tomkeuper) [reply versionsupport_v1_20_r5/src/main/java/com/tomkeuper/bedwars/support/version/v1_20_R5/v1_20_R5.java]: Would you be able to test? if not Ill see if i can at some point.

Comment (Lagggpixel) [reply versionsupport_v1_20_r5/src/main/java/com/tomkeuper/bedwars/support/version/v1_20_R5/v1_20_R5.java]: I've heard from other minigames that this still occasionally causes issues. I probably won't be free to test.

File: versionsupport_v1_20_r5/src/main/java/com/tomkeuper/bedwars/support/version/v1_20_R5/v1_20_R5.java (hunk 2 above)
Comment (Lagggpixel) [reply versionsupport_v1_20_r5/src/main/java/com/tomkeuper/bedwars/support/version/v1_20_R5/v1_20_R5.java]: I think I might've tagged the wrong snippet.

File: versionsupport_v1_20_r5/src/main/java/com/tomkeuper/bedwars/support/version/v1_20_R5/v1_20_R5.java (hunk 5 above)
Comment (Lagggpixel) [reply versionsupport_v1_20_r5/src/main/java/com/tomkeuper/bedwars/support/version/v1_20_R5/v1_20_R5.java]: True

**Review by tomkeuper — APPROVED** [review]
